TEE.md

```diff
@@ -0,0 +1,77 @@
+
+This is a good moment to remind the reader that threshold cryptography requires threshold operational nodes. Anything less than that, and the protocol stalls, in which case, no funds can be moved.
+
+Therefore, Near-One will roll-out its TEE implementation in two phases:
+- Soft Launch: All mainnet nodes are running within TEEs. Their key shares are backed-up outside of the TEE.
+- Hard Launch: All mainnet nodes are running within TEEs. Their key shares are backed-up inside a different TEE.
+
+In order to protect the network from the worst case scenario (complete loss of funds due to loss of key shares or an incapacitated network), a disaster recovery plan is prepared.
+
+
+## Disaster Recovery
+Disaster recovery is a plan intended to prevent a permanent loss of the signing quorum.
+As long as the secret shares of the MPC nodes are securely backed-up outside of the TEE environment in which the node is running, it is highly likely that the network will be able to recover from otherwise catastrophic events.
+
+Therefore, the disaster recovery plan aims at establishing a secure mechanism allowing to back up the secret shares held by the nodes. It encompasses two steps:
+1. (Back-up Generation): export the private keys from the TEE in a secure manner.
+2. (Recovery): when required, securely import the private keys into the node.
+
+### Back-up mechanism:
+At a high-level, the back up mechanism works like this:
+
+1. The node running inside the TEE and the backup recipient must agree on a key used to encrypt the secrets.
+2. The node encrypts the sensitive secret shares with the derived key.
+3. The node exports the encrypted secret shares.
+
+#### Agreeing on an encryption key
+A pair-wise key establishment scheme can be leveraged to establish a symmetric key. A suitable implementation can be found in [libsodium](https://doc.libsodium.org/) and its rust wrapper [libsodium-rs](https://docs.rs/crate/libsodium-rs/latest). Specifically, the **key exchange protocol based on X25519** would be a good fit for the given set-up (c.f. [libsodium](https://doc.libsodium.org/key_exchange) and [libsodium-rs](https://docs.rs/libsodium-rs/latest/libsodium_rs/crypto_kx/index.html)).
+
+_Note: This key exchange protocol is on Curve25519, which is different from the Edwards curve used by NEAR._
+
+The key establishment scheme requires that the node as well as the backup recipient have *mutually authenticated Curve25519 public keys*. The NEAR blockchain can be leveraged for this, more specifically:
+- the node operator whitelists a NEAR-account associated to the backup recipient in the MPC smart contract.
+- the backup recipient submits a public key (Curve25519) to the MPC smart contract. The smart contract only accepts it if the account has been whitelisted by the node operator.
```

Comment:
How do we make sure the public key on the contract is always up to date? Is it possible to end up in a situation where the backup provider or recipient generates a new public key (e.g., after a restart), so the backup is saved but can no longer be decrypted because it was encrypted using keys that are no longer accessible?

Reply:
The key used for encrypting the secret at rest is different from the key used for encrypting the secret in transit.